### ISSUE TITLE
lz4 -t always succeeds on any file

### ISSUE THREAD

**Issue by nrgiii** [lz4/lz4#199]
Running lz4 -t on any readable file reports success and exits with status 0.  This is unexpected.  I think lz4 should complain if it is passed a file that is not in lz4 format.  The -t option is supposed to "test compressed file integrity" per the help screen, however it does no such test AFAIKT.

normg@bunk>lz4 -t /usr/bin/ssh
Successfully decoded 429936 bytes
normg@bunk>ls -l /usr/bin/ssh
-rwxr-xr-x 1 root root 429936 Aug 17  2015 /usr/bin/ssh

Norm Green

**Comment by t-mat** [comment]
Hi @nrgiii,

I've confirmed it and I agree with you.  `lz4 -t` should show proper message and/or exit code.

@Cyan4973, it seems there are some problems:

1. Single file decompression path in `main()` [ignores return value of `DEFAULT_DECOMPRESSOR()`](https://github.com/Cyan4973/lz4/blob/8b562a4f37c33ee026bc5976c7cff9c210c5d35f/programs/lz4cli.c#L564).  Therefore `operationResult` (exit code) is always 0.
   - Also return value of [`DEFAULT_COMPRESSOR()` is ignored](https://github.com/Cyan4973/lz4/blob/8b562a4f37c33ee026bc5976c7cff9c210c5d35f/programs/lz4cli.c#L579).
2. Since [`case 't'` calls `LZ4IO_setOverwrite(1)`](https://github.com/Cyan4973/lz4/blob/8b562a4f37c33ee026bc5976c7cff9c210c5d35f/programs/lz4cli.c#L391) (`g_overwrite` = 1), `selectDecoder()` in lz4io.c doesn't fail even [it recognizes wrong header magic number](https://github.com/Cyan4973/lz4/blob/8b562a4f37c33ee026bc5976c7cff9c210c5d35f/programs/lz4io.c#L945).  It always goes to `LZ4IO_passThrough()`.

Basically, these behaviours have been [introduced with pass-through mode](https://github.com/Cyan4973/lz4/commit/e7fb4d1#diff-abb30b1425dc9e3e5441d294b2f7e909L704).  I think possible solutions are

1. Add extra mode for LZ4IO which overrides pass-through mode.  Something like `LZ4IO_setIntegrityCheckMode()`.
2. Redefine proper behaviour of pass-through mode.

**Comment by Cyan4973** [comment]
Thanks for notification @nrgiii and for clear debugging @t-mat .

This is indeed a problem.
I guess we need a way to disable pass-through mode when `-t` is selected.

**Comment by t-mat** [comment]
@Cyan4973,  I see.  So will it be look like below?

 - `PATCH` indicates difference
 - `EXM_THROW(44, "Unrecognized header ..."):` seems problematic.

```C
// lz4io.c
// PATCH : As default, pass-through mode is enabled
static int g_passThrough = 1;

// PATCH : Set pass-through mode
int LZ4IO_setPassThrough(int enable) {
    g_passThrough = (enable != 0);
    return g_passThrough;
}

static unsigned long long selectDecoder(dRess_t ress, FILE* finput, FILE* foutput) {
    ...
    default:
        if (nbCalls == 1) {   /* just started */
            // PATCH : Add g_passThrough for explicit pass-through mode.
            if (g_passThrough && g_overwrite)
                return LZ4IO_passThrough(finput, foutput, MNstore);
            //
            // PATCH?
            // Here, should we return error code instead of EXM_THROW()/exit()?
            //
            EXM_THROW(44,"Unrecognized header : file cannot be decoded");
        }
    ...
}

// lz4cli.c
int main(int argc, char** argv) {
    ...
        /* Test integrity */
        case 't':
                // PATCH : Disable pass-through mode
                LZ4IO_setPassThrough(0);
                decode=1; LZ4IO_setOverwrite(1); output_filename=nulmark; break;
    ...
}
```

**Comment by Cyan4973** [comment]
Indeed, it's an explicit way to do it, and it would work.

I was wondering about another possibility, although I need to look deeper into the code to see if it can work : 

"pass-through" mode is supposed to be only enabled in combination with `-f` command : 

```
cat anyfile | gzip -df > /dev/null
echo $?
0
cat anyfile | gzip -d > /dev/null
gzip: stdin: not in gzip format
echo $?
1
```

This is imitated by `lz4` (with minor differences) : 

```
cat anyfile | lz4 -df > /dev/null
echo $?
0
cat anyfile | lz4 -d > /dev/null
Error 44 : Unrecognized header : file cannot be decoded
echo $?
44
```

which means that, "pass-through" mode is not automatic, it's only enabled by `-f`.

So, the basic concept is, `-t` should be have the same as `-d -c > /dev/null`.
Which means that, in the best case, there is no need to add another explicit setting.
But since it doesn't work like this currently, it's unclear to me if what I just described is sensible or not.

**Comment by Cyan4973** [comment]
Got it,
`-t` also enables `setOverwrite`.
That's why it behaves like "pass-through".

Also it needs to be confirmed, I believe the problem is that, without it, `fileio` will ask for a confirmation prompt when writing to `/dev/null` because it already exists.
So we need a way to not trigger such confirmation, and still behave as if `-f` is not enabled.

Maybe special-casing `nulmark` will be enough ...

**Comment by nrgiii** [comment]
The patch code from t-mat above works for me.  With the patch, lz4 -t on a non-lz4 file now prints an error message and exits with a non-zero status code.

normg@bunk>./lz4 -t /usr/bin/ssh
Error 44 : Unrecognized header : file cannot be decoded

This is sufficient for what I need.

thanks,

Norm
